### PR TITLE
Add notifications for approved & rejected suggestions

### LINKED ISSUE #2094: Notify suggestion authors when their suggestions get accepted/rejected
*This issue was created automatically by a [script](https://github.com/mathjazz/bugzilla2github/).*
## [Bug 1398862](https://bugzilla.mozilla.org/show_bug.cgi?id=1398862)
Bug Reporter: @mathjazz
CC:   @flodolo

We should probably do that periodically and aggregate notifications.

### PR DESCRIPTION
**Pull request by eemeli**
Closes #2094 

Adds a script for sending notifications on approved & rejected suggestions to their authors; this should be run every 24 hours by a cron job:

```sh
./manage.py send_review_notifications
```

This follows the same structure and pattern established for suggestion notifications as used in #2329 and #2330.

The message formatting is done all in Python, as it effectively uses a two-variable top level case selector that isn't really supported by current languages; MF2 will deal with it fine.

### CODE REVIEW

File: pontoon/projects/management/commands/send_approved_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_approved_notifications.py:45]: I would simplify the logic and only list project-locales with reviewed strings, same as we do in the unreviewed suggestions notifications.

Count is useful, I agree, but I'd rather not mess with the sentences, and instead use a graphic, e.g. Screenshot 2022-03-15 at 19 49 32 

Maybe do that in a follow-up, also for unreviewed suggestion notifications?

**Comment by eemeli** [comment]
Renamed as "review notifications" and applied other suggestions from @mathjazz.

File: pontoon/projects/management/commands/send_approved_notifications.py
Comment (eemeli) [reply pontoon/projects/management/commands/send_approved_notifications.py]: The sentence construction isn't really complicated, it's just complex.

I think this and the discussion on due-date notifications is sounding like an impending review of all notifications.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_review_notifications.py:72]: All approved translations must be active, so we can remove this condition.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_review_notifications.py:81]: I'm not sure about this one. I see why it could make sense when you add a suggestion to an existing approved translation and that suggestion gets accepted. But if I make two suggestions, and one gets approved and the other one rejected, that should count as 1 rejected and 1 approved.

File: pontoon/projects/management/commands/send_approved_notifications.py
Comment (mathjazz) [reply pontoon/projects/management/commands/send_approved_notifications.py]: For unreviewed suggestions notifications we also send just one notification per user, here we send one per project-locale combination. (And add counts.)

I'm also curious what @flodolo thinks about the message structure.

Comment (flodolo) [reply pontoon/projects/management/commands/send_approved_notifications.py]: > I'm also curious what @flodolo thinks about the message structure.

It feels very complex to build. As the recipient of such message, it also forces my brain to do some extra work, because the structure changes depending on the data, and it's not as immediate to parse.

I'd be happy with something minimal (and less elegant/natural) like.

```
Some of your suggestions were reviewed:
* Approved: XX strings
* Rejected: XX strings
```

And drop approved/rejected if the number is 0.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (eemeli) [reply pontoon/projects/management/commands/send_review_notifications.py:81]: If you make two suggestions for the same entity and one of them is approved, I would think that the context makes it kinda obvious what's happened?

One alternative that I considered was counting only rejections of suggestions that were never approved, but there doesn't appear to be a way of filtering for only those.

File: pontoon/projects/management/commands/send_approved_notifications.py
Comment (mathjazz) [reply pontoon/projects/management/commands/send_approved_notifications.py]: Thanks, Flod. What is your take on 1 notification per user vs one pre each project-locale-user combination?

Comment (flodolo) [reply pontoon/projects/management/commands/send_approved_notifications.py]: I think 1 notification would be ideal, as long as we can link to specific strings for each project in the message.

I also don't expect notifications to include a lot of projects, so maybe the end result is the same in most cases (for most users, it will be about 1 project).

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [reply pontoon/projects/management/commands/send_review_notifications.py:81]: Agreed with both points.

File: pontoon/projects/management/commands/send_approved_notifications.py
Comment (eemeli) [reply pontoon/projects/management/commands/send_approved_notifications.py]: As it would appear from this discussion that the spec for the notification contents that I proposed in https://github.com/mozilla/pontoon/issues/2094#issuecomment-1066754638 isn't good enough after all, how should we proceed with re-specifying it?

Comment (mathjazz) [reply pontoon/projects/management/commands/send_approved_notifications.py]: I suggest we follow the lead of Unreviewed suggestions notifications and send only 1 notification per user with the following content:

```
Your suggestions in the following projects have been reviewed:
* Firefox (X Approved, Y Rejected)
* Common Voice (Z Approved)
```

I would also implement a new `review_time` filter, similar to the `time` filter, and use it in the link.

In a followup, we could also add count to Unreviewed suggestions notifications.

Comment (eemeli) [reply pontoon/projects/management/commands/send_approved_notifications.py]: Okay; is this how the notification would look like for a single approved string in one project?

```
Your suggestion in the following project has been reviewed:
* Firefox (1 Approved)
```

That seems a bit clumsy. Would it make sense to use a different format if all the strings are in one project, like this?

```
Your suggestions in Firefox have been reviewed (2 Approved, 1 Rejected)
```

And with a special case for a single string:

```
Your suggestion in Firefox has been Approved
```

Also, if we're listing projects and not project+locale combos, how should we create the links if a user had strings reviewed for one project in more than one language?

Comment (flodolo) [reply pontoon/projects/management/commands/send_approved_notifications.py]: Personally, I prefer consistent message structure for readability (I can quickly scan the relevant info, which is project and number of strings review).

Review for multiple locales+projects within two cron job runs is an edge case, but we clearly need to deal with it somehow. Would it be easier to send separate notifications for each locale?

Comment (mathjazz) [reply pontoon/projects/management/commands/send_approved_notifications.py]: Agreed with Flod about the message structure.

For Unreviewed, if more than one locales are affected, we list project for each locale and set the URL accordingly. We should show the locale identifier, as well. For example:

```
Your suggestions in the following projects have been reviewed:
* Firefox (Slovenian · sl): X Approved, Y Rejected
* Common Voice (french · fr): Z Approved
```

Comment (eemeli) [reply pontoon/projects/management/commands/send_approved_notifications.py]: Notification contents refactored as per above. Left out "in the following projects" as redundant, and leading to complexity in needing to match the plural case to the actual count of projects.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_review_notifications.py:45]: Let's use HTML lists here, which gives as more styling powers and structure. It might be easier to do with a template, see `suggestion_notification.jinja`.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_review_notifications.py:33]: I would suggest we filter by author in all cases. Or block this whole patch on #2049?

Comment (eemeli) [reply pontoon/projects/management/commands/send_review_notifications.py]: Added author filter; that does make sense always. Still including the `&string=` when there's only one.

Recommend against blocking on #2049, because the solution to that is likely to be built on top of #2452, and that would require that whole stack of things to resolve before we get to merge this.

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (mathjazz) [inline pontoon/projects/management/commands/send_review_notifications.py:44]: After testing this layout a bit, I propose we make two changes:
1. Drop Locale name and only leave locale code in the brackets. That should make notification data easier to digest and more condensed. Also, some locale names are long. As in 4 words long.
2. Use `PROJECT (LOCALE)` piece as anchor instead of the `msg` piece.

**Comment by mathjazz** [comment]
Other than the proposal in the comment above, I think this patch is good to go!

File: pontoon/projects/management/commands/send_review_notifications.py
Comment (eemeli) [reply pontoon/projects/management/commands/send_review_notifications.py:44]:
```suggestion
            desc += f'\n<li><a href="{url}">{project.name} ({locale.code})</a>: {msg}</li>'
```

**Comment by mathjazz** [comment]
Fix black and we're good!

**Review by mathjazz — APPROVED** [review]
Excellent work!

**Pull request merged**
